model-optimizer/mo/graph/connection.py

```diff
@@ -83,15 +83,45 @@ def set_source(self, port):
 
         if self.graph.stage == 'front':
             scr_node = port.node
+
+            source_fw_names = []
+            for dst_port in port.get_connection().destinations:
+                edge_attrs, u, v, key = dst_port.get_in_edge_attrs(data=True)
+                for attr in edge_attrs:
+                    if attr == "fw_tensor_debug_info":
+                        source_fw_names += edge_attrs[attr]
+            # remove duplicates
+            source_fw_names = list(set(source_fw_names))
+
             # Reconnecting all destinations as consumers to the source port preserving edge attrs
             for dst_port in self.destinations:
                 edge_attrs, u, v, key = dst_port.get_in_edge_attrs(data=True)
                 if u is not None:
                     edge_attrs['out'] = port.idx
+                    attr = "fw_tensor_debug_info"
+                    if attr in edge_attrs:
+                        if attributes_save_mode == "merge":
+                            # Merge attributes from old source edges
+                            edge_attrs[attr] += source_fw_names
+                        elif attributes_save_mode == "source":
+                            # Copy attributes from old source edges
+                            if len(source_fw_names) > 0:
+                                edge_attrs[attr] = source_fw_names
+                            else:
+                                del edge_attrs[attr]
+                        # in case of dest mode attributes on edges are not changed
+                    else:
+                        if attributes_save_mode != "dest":
+                            if len(source_fw_names) > 0:
+                                edge_attrs[attr] = source_fw_names
+
```

Comment:
Consider moving this to a separate function to make function shorter. The same comment for other blocks of code you added. These new function should not be a public API but some internal module functions